File: Solutions/java/AdventOfCode.java
package Solutions.java;

public class AdventOfCode {
    public static void main(String[] args) {
        if(args.length == 0) {
            System.out.println("Please provide a date for which to solve a challenge");
            return;
        }

        var date = getDate(args[0]);

        var inputFile = getInputFile(args, date);
        solve(date, inputFile);
    }

    public static Date getDate(String dateStr) {
        var splittedDate = dateStr.split("/");
        return new Date(Integer.parseInt(splittedDate[0]), Integer.parseInt(splittedDate[1]));
    }

    public static void solve(Date date, String inputFile) {
        var className = String.format("Solutions.java.y%d.Solution%02d", date.year, date.day);
        try
        {
            var classObject = Class.forName(className);
            var instance = classObject.getDeclaredConstructor().newInstance();
            
            var part1 = classObject.getMethod("Part1", String.class);
            var part2 = classObject.getMethod("Part2", String.class);
            
            part1.invoke(instance, inputFile);
            part2.invoke(instance, inputFile);
        }
        catch(ClassNotFoundException e)
        {
            System.out.printf("Solution not found for %d/%02d\n", date.year, date.day);
        }
        catch(Exception e)
        {
            System.out.println("Something went wrong");
        }
    }

    public static String getInputFile(String[] args, Date date) {
        if (args.length >= 2)
            return args[1]; 
        else 
            return String.format("Inputs/%d/%02d.txt", date.year, date.day);
    }
}

File: Solutions/kotlin/adventofcode.kt
package Solutions.kotlin;

import kotlin.reflect.cast

fun main(args: Array<String>) {
    if(dateProvided(args)) {
        println("Please provide a date for which to solve a challenge")
        return
    }

    val date = getDate(args)
    
    val solution = getSolution(date)
    val inputFile = getInputFile(date, args)

    solve(solution, inputFile)
}

fun dateProvided(args: Array<String>): Boolean {
    return args.size == 0
}

fun getDate(args: Array<String>): Date {
    val splittedDate = args[0].split("/")
    return Date(splittedDate[0].toInt(), splittedDate[1].toInt())
}

fun getSolution(date: Date): Solution {
    val className = getClassName(date)

    val classObject = Class.forName(className).kotlin
    val ctor = classObject.constructors.first()
    val obj = ctor.call() as Solution

    return obj
}

fun getClassName(date: Date): String {
    return "Solutions.kotlin.y%d.Solution%02d".format(date.year, date.day)
}

fun getInputFile(date: Date, args: Array<String>): String {
    if (args.size >= 2)
        return args[1] 
    else 
        return "Inputs/%d/%02d.txt".format(date.year, date.day)
}

fun solve(solution: Solution, inputFile: String) {
    solution.Part1(inputFile);
    solution.Part2(inputFile);
}